### DISCUSSION TITLE
❓  openapi schema can auto-convert z.date()  -> zChanged.dateOut()

### DISCUSSION THREAD

**Discussion by HardCoreQual** [RobinTail/express-zod-api#537] [Ideas]
now z.date() from output schema in during generete OpenApi.response  will throw error  `Zod type ZodDate is unsupported`

for example prisma ORM have zod generator that will create z.date(), but now it don't can be used

Think for z.dateIn() will require more complex logic that will rewrite each  enpoint.input for replace date for make that it work in openapi-generator and in endpoint.input

**Comment by RobinTail** [comment]
Hello @HardCoreQual 

`z.date()` represents JS `Date` object.
see https://github.com/colinhacks/zod#dates

Your endpoint can not receive `Date` as an input, since the data comes either in Request's query (which contains strings) or body (which is JSON, that does not support `Date`).

Therefore, the way to transmit the dates is by converting them to strings.

However, there is an incoming and an outgoing data flow in your Endpoint.
For the incoming you'd need to convert string into Date, while for outgoing you might need to do the opposite: convert Date to string, in order to make it transmittable.

That's why I made `z.dateIn()` and `z.dateOut()`. More details here:
https://github.com/RobinTail/express-zod-api#dealing-with-dates

You say that you're using some generator (Prisma ORM) to create zod schemas. 
If it produces `z.date()` for input — this is a mistake, it just wouldn't work without preprocessing the incoming string into a Date, as described in the original zod's documentation.

So, I believe the generator you mentioned should either be improved or you'd have to adjust its results, or you can write zod schema yourself.

**Reply by HardCoreQual** [reply]
I'm not sure that I understand because is imposible to do 

for input  -- recursive parse schema and convert each z.date() -> z.dateIn()
for ouput -- recursive parse schema and convert ecah z.date() -> z.dateOut()

**Reply by RobinTail** [reply]
we were discussing some generator, right?
I presume it generates a text, or a file that contains text.
You can:
- write an utility that performs changes in the text in order to fix the issue with dates;
- fork the generator and fix the cause of the issue (input can not be Date, output too by the way, they are always strings);
- avoid using this generator and write zod schemas manually

**Comment by HardCoreQual** [comment]
in standard situations z.date() work fine for OUTPUT because `JSON.stringify({ date: new Date() })` will convert it to string, but it will create bug in during generate openapi, I just think that autoconversion in OpenApi will not broke anything because now it throw error

**Comment by RobinTail** [comment]
Yes, in case of the output, `Date` is automatically converted to the string while converting the whole response object to JSON.
However, this does not work well when it comes to the frontend types.
If your endpoint outputs `z.date()` it means that the frontend should expect `Date`, while it has been silently converted to a string, not Date. It leads to confusion and discrepancy.
Therefore, I intentionally excluded `z.date()` from the OpenAPI generator, in order to draw your attention to this problem, so you'd choose the right type `z.dateIn()` or `z.dateOut()` keeping in mind that Dates are always a subject to conversion.
@HardCoreQual

**Comment by RobinTail** [comment]
Here is an interesting thing regarding incoming date problem.
JSON body parser `express.json()` has the `reviver` property that could be utilized for restoring `Date` from `string`:
https://expressjs.com/en/5x/api.html#express.json

There is a however a question: how to detect such string reliably and fast.
